jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py

```diff
@@ -90,106 +131,134 @@ def __init__(self, *args, **kwargs):
         self.session.hub.subscribe(self, SubsetUpdateMessage, handler=self._on_subset_update)
         self.session.hub.subscribe(self, LinkUpdatedMessage, handler=self._on_link_update)
 
+    @property
+    def user_api(self):
+        return PluginUserApi(self, expose=('multiselect', 'dataset', 'aperture',
+                                           'background', 'background_value',
+                                           'pixel_area', 'counts_factor', 'flux_scaling',
+                                           'calculate_photometry',
+                                           'unpack_batch_options', 'calculate_batch_photometry'))

```

Comment:
I think it is too confusing to have 2 different ways to call "calculate photometry". The method should be smart enough to know if multiple things are selected or not and then behave accordingly.

```suggestion
                                           'calculate_photometry',
                                           'unpack_batch_options'))
```

Similarly, I am also not sure about exposing things like "aperture" and so on that can conflict with the batch (multiselect) mode, unless those are read-only?

Should `'unpack_batch_options'` even be exposed at public? I vaguely remember discussions about making the API available but not public yet until we can figure out more stuff? Has that changed?

Reply:
> I think it is too confusing to have 2 different ways to call "calculate photometry". The method should be smart enough to know if multiple things are selected or not and then behave accordingly.

Internally the button press goes to the same method, but the arguments and API docs for the two when you are _overriding_ plugin values are different.  We could maybe have them intelligently redirect to the other when there aren't any overrides (calling `calculate_photometry()` when `plugin.multiselect is True` could just redirect to the batch mode instead of the current behavior of raising an error).  Or if anyone has suggestions for consolidating the API, I'm open to ideas!

> Similarly, I am also not sure about exposing things like "aperture" and so on that can conflict with the batch (multiselect) mode, unless those are read-only?

I can't think of a place where there would be a conflict, can you show an example?

> Should 'unpack_batch_options' even be exposed at public? I vaguely remember discussions about making the API available but not public yet until we can figure out more stuff? Has that changed?

This is the last ticket in the effort, so this (and the review) is figuring out the stuff.  

`unpack_batch_options` was _originally_ intended to remain private, but @eteq requested the ability to expose the list of options generated from the UI so they could then be modified and manually passed through the API.  Making this public (and implementing the case of passing no input arguments) accomplishes that, as described in the PR description.

If we do want to hide things for a release cycle to iterate further internally without having to worry about backwards compatibility, that is definitely an option we can consider.

Reply:
>  Or if anyone has suggestions for consolidating the API, I'm open to ideas!

```python
def calculate_batch_photometry(self, options=[], add_to_table=True, update_plots=True,
                               full_exceptions=False):
```

vs

```python
def calculate_photometry(self, dataset=None, aperture=None, background=None,
                         background_value=None, pixel_area=None, counts_factor=None,
                         flux_scaling=None, add_to_table=True, update_plots=True):
```

Since neither of these methods where public before, we are free to change the API as we please here. We can even make both of these internal, i.e., `_calculate_batch_photometry` and `_calculate_photometry`. Then we can create a new public `calculate_photometry` that can even have no inputs. It would act based on traitlet values and call `_calculate_batch_photometry` or `_calculate_photometry` with preset input values.

What do you think?

> the ability to expose the list of options generated from the UI so they could then be modified and manually passed through the API

I still cannot quite make sense of a science use case for this. I don't want us to go out of the way to implement something that no one ends up using in real life but only looks good as a thought exercise. Maybe a concept notebook to illustrate?

Reply:
> Then we can create a new public calculate_photometry that can even have no inputs.

I'd be ok with turning the current `calculate_photometry` -> `_calculate_photometry` and then have a public method with no arguments for the single mode.  But half the point of the batch mode is to enable these advanced workflows that are not accessible from the UI, so that at least needs to be able to accept `options` (or we flip it and let all kwargs take lists and just enforce they are the same length, but then that becomes more difficult in certain workflows, like the one described next).

> I still cannot quite make sense of a science use case for this.

Here is a simple version of the case that (if I understood correctly) was requested ([also described here](https://github.com/spacetelescope/jdaviz/pull/2465#issuecomment-1746948824))


https://github.com/spacetelescope/jdaviz/assets/877591/e2b36f6b-7f5a-4f47-9e2d-2f9beda7587f

Reply:
I was under the impression that plugin API is to allow people to interact with plugin as it is designed, but via API.

However, in that video above, you are doing something in the API that is technically impossible to do in plugin; i.e., selecting multiple background values in multi-select mode. It breaks feature parity between GUI and API. I don't know how to feel about it.

Reply:
I think it's fine if the API exposes advanced functionality not available through the UI, but the _designs_ should be analogous to each other so that transitioning and writing scripts is as intuitive as possible.  Do you have any alternate suggestions to achieve the requirements?  This was in the original plan for this work (exact syntax aside).

Reply:
It is a headache because if user start providing Subset names in the background section in the `dict`, they are bypassing the UI check. It will be very error prone. Are you limiting this special API access to manually provided scalar values? Even then, unit is *assumed* in the plugin, you cannot provide a Quantity (because GUI does not allow for it).

We're just making it easier for people to make mistakes.

Reply:
I would argue that we need to move/copy checks into the python side then (including forbidding quantity objects if that's necessary).

Reply:
The API is now no-longer public, so we can use these methods internally and iterate as necessary.  Once we're happy, we can then enable making these public.